src/binary-reader.cc

```diff
@@ -3041,7 +3041,8 @@ Result BinaryReader::ReadModule(const ReadModuleOptions& options) {
                "function signature count != function body count");
   // This is checked in ReadDataSection, but it must be checked at the end too,
   // in case the data section was omitted.
-  ERROR_IF(num_data_segments_ == 0 && data_count_ != kInvalidIndex,
+  ERROR_IF(num_data_segments_ == 0 && data_count_ != kInvalidIndex &&
+               data_count_ != 0,
```

Comment:
Hmm.. I wonder how we missed this case in #2432?  @SoniEx2 ?

Reply:
This test (asserting that a zero data count section without a data section is valid) wasn't part of that PR and test (https://github.com/WebAssembly/spec/pull/1801). It was added as a direct commit right afterwards (https://github.com/WebAssembly/spec/commit/7ff01f577b3cb0695bddd9083b7d09a289a3ea21#diff-63134a480c82fef0f49d50176aedf7cb823b5f0c187a4a3663165e2affbf8aedR486). So I don't think we tested for that situation at the time.

Reply:
ah yeah, we made a mistake when trying to "simplify" this condition! guess there wasn't a test for it.